### PR TITLE
rsx: Surface format remapping enhancements

### PR DESCRIPTION
**Pull request by kd-11**
TODO:
- [x] Confirm XBGR8 behaviour on realhw
- [x] NVIDIA blur issue
- [x] Fix float16 issues

N/B: Colormask for XBGR is not subject to swapping - verified with hw test.

Fixes https://github.com/RPCS3/rpcs3/issues/8902

### CODE REVIEW

**Comment by MarioSonic2987** [comment]
Fixes broken lighting in Watch_Dogs. Excessive blur is still here.

**Before**:
image

**After**:
image

File: rpcs3/Emu/RSX/GL/GLDraw.cpp [hunk 1]
```diff
@@ -144,10 +144,14 @@ void GLGSRender::update_draw_state()
...
 		bool color_mask_a = rsx::method_registers.color_mask_a(index);
 
-		if (rsx::method_registers.surface_color() == rsx::surface_color_format::g8b8)
+		switch (rsx::method_registers.surface_color())
 		{
-			//Map GB components onto RG
+		case rsx::surface_color_format::b8:
+			rsx::get_b8_colormask(color_mask_r, color_mask_g, color_mask_b, color_mask_a);
+			break;
+		case rsx::surface_color_format::g8b8:
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLDraw.cpp:152]: Needs default case to prevent warning

File: rpcs3/Emu/RSX/VK/VKGSRender.cpp [hunk 2]
```diff
@@ -1646,8 +1653,15 @@ bool VKGSRender::load_program()
 		bool color_mask_r = rsx::method_registers.color_mask_r(index);
 		bool color_mask_a = rsx::method_registers.color_mask_a(index);
 
-		if (rsx::method_registers.surface_color() == rsx::surface_color_format::g8b8)
+		switch (rsx::method_registers.surface_color())
+		{
+		case rsx::surface_color_format::b8:
+			rsx::get_b8_colormask(color_mask_r, color_mask_g, color_mask_b, color_mask_a);
+			break;
+		case rsx::surface_color_format::g8b8:
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKGSRender.cpp:1661]: Same here

**Comment by kd-11** [comment]
Check if blurriness is fixed.

**Comment by MarioSonic2987** [comment]
Blur is fixed.
image

**Comment by Darkhost1999** [comment]
Was going to see if Blur was fixed in Uncharted 3 because in master it looks like this without Relaxed zcull. Relaxed zcull fixes this specific area of act 9.
Screenshot 2022-01-15 174430

But what may interest you is:
Uncharted
All of these Fatal errors 
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/7876106/RPCS3.log)

**Comment by Xcedf** [comment]
@Darkhost1999 the issue that you showed is nothing to do with this PR, Motion Blur in Uncharted 3/TLoU never worked properly on rpcs3, for some reason it interferes with depth buffer, as a result smth like this

and Relaxed ZCull also fixes nothing
Approx ZCull

Relaxed 

Enable Motion Blur patch if you want to rid of that thing and don't use Precise ZCull it may actually make this thing worse.

**Comment by Darkhost1999** [comment]
How are you not getting fatal errors when trying to boot the game on the PR though?

**Comment by Xcedf** [comment]
@Darkhost1999 I don't know, just not happening, try to remove your shader cache.

**Comment by Darkhost1999** [comment]
After File - all titles - remove shader cache
Games that have regressed in this PR
Angry birds Star wars 
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/7876319/RPCS3.log)
Battlefield 3 (Not the same error Just a freeze no fatal error)
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/7876328/RPCS3.log)
Call of Duty Ghosts
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/7876334/RPCS3.log)
Dante's Inferno
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/7876336/RPCS3.log)

I have more. But yes I'm aware the Blur issue in Uncharted 3 is not related to this PR.
Whereas the fatal error is definitely related to this PR

**Comment by coolllman** [comment]
Call of duty not work with this pr, when start stage result black screen, on master 13185 all good

**Comment by RainbowCookie32** [comment]
13185 is the latest build, so if it's working now then it isn't really an issue?

**Comment by coolllman** [comment]
In this pr issue, on master all work

**Comment by RainbowCookie32** [comment]
> In this pr issue, on master all work

Then providing some info would be nice, like a screenshot of the issue and a log

**Comment by coolllman** [comment]
Freeze on cache load
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/7876441/RPCS3.log)
Delete all caches, boot, but freeze on start stage
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/7876442/RPCS3.log)

**Comment by coolllman** [comment]
Battlefield Bad Company 2 the same freeze on start

**Comment by coolllman** [comment]
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/7876445/RPCS3.log)

**Comment by coolllman** [comment]
I think more games affected

**Comment by SlyCooperReloadCoded** [comment]
I'm testing the PR now, Watch Dogs fails to boot with the same fatal errors as Darkhost1999 is getting.  Seems it's not an isolated instance.

**Comment by SlyCooperReloadCoded** [comment]
Oh, and here's a log for that:
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/7876469/RPCS3.log)

**Comment by coolllman** [comment]
With disable native float16 all work

**Comment by SlyCooperReloadCoded** [comment]
Wow, Disable Native Float16 Support does indeed fix the fatal errors, however I have a feeling this wasn't intended.

**Comment by kd-11** [comment]
Please retest if float16 bug is fixed.

**Comment by coolllman** [comment]
Float16 bug fixed, but fps drop from 60 to 33 in cod world war and maybe in other games(((

**Comment by kd-11** [comment]
I'm confident the fps drop is not caused by this PR. It is more likely something else was merged into master between when you last tested master and when you tested this out. Nothing much has been changed here other than some GPU instructions getting added. Unless you're at 99% GPU load (which is unlikely unless you're using intel integrated graphics) then nothing should be different with this PR.

**Comment by coolllman** [comment]
Infamous 2  15fps drop too, all games huge fps drop, preloading cache and compiling cache now very long

**Comment by coolllman** [comment]
No, nothing merge

**Comment by coolllman** [comment]
Latest merge 13 hours ago

**Comment by coolllman** [comment]
I test on latest master,your pr with float16 bug and your fixed pr

**Comment by kd-11** [comment]
I'll need other people to corroborate this as I cannot see any difference on my system.

**Comment by Xcedf** [comment]
Well I have a small drop in Infamous 2, I bet it goes to those rounding routines, but definitely not a 15fps hit.
master

pr

**Comment by Darkhost1999** [comment]
Don't got Infamous 2 but I got Infamous
Sorry my spawn area is right next to some vents and I'm using remote desktop. But the point is FPS. Not visual quality.
Master

PR

Aka within the margin of error no perceived difference.
Oh and float16 fatal errors are gone.

**Comment by coolllman** [comment]
Update driver to 5.11.23 reboot pc and fps back, but on the 3 boot fps drop again, run master fps good, reboot pc fps back, boot game 5 times - fps drop and only reboot fix it, on master all works flawestly

**Comment by kd-11** [comment]
> Update driver to 5.11.23 reboot pc and fps back, but on the 3 boot fps drop again, run master fps good, reboot pc fps back, boot game 5 times - fps drop and only reboot fix it, on master all works flawestly

This is a known windows bug. It can happen with any rpcs3 version. Nobody knows what causes it and it never seems to occur on any dev machines that I am aware of.

**Comment by coolllman** [comment]
This first time i see this bug and only on pr, i try 15 times boot on master and never see this bug

**Comment by kd-11** [comment]
> Strange, this first time i see this bug and only on pr, i try 15 times boot on master and never see this bug, when fps drop i try boot again many times and on pr fps low, i run master and fps is back, run again pr fps low. Try anyone multiple boots various games, is fps drop?

No, I mean, I haven't rebooted my machine despite starting rpcs3 dozens of times and performance is the same. This issue has been around for a while, we never figured out why some people have it and not others.

@Xcedf I have altered the rounding behaviour slightly, check if it restores the 1-2fps difference observed.
We also need retesting of the original watchdogs bug, just to make sure I didn't accidently undo the fix.

**Comment by kd-11** [comment]
Actually @coolllman the huge perf drop is because windows flagged rpcs3 as an unstable application and enabled some flags. See https://docs.microsoft.com/en-us/windows/win32/win7appqual/fault-tolerant-heap#:~:text=To%20disable%20Fault%20Tolerant%20Heap,longer%20activate%20for%20new%20applications.

**Comment by Xcedf** [comment]
@kd-11 yes, have my 2fps back

**Comment by MarioSonic2987** [comment]
Watch_Dogs bug is still fixed.
image

**Comment by coolllman** [comment]
Thank you kd-11, i will read about unstable application, in new pr all good, i change name exe

**Comment by SlyCooperReloadCoded** [comment]
float16 support now works, however I'm now getting device lost fatal errors during gameplay.

**Comment by lextra2** [comment]
> No, I mean, I haven't rebooted my machine despite starting rpcs3 dozens of times and performance is the same. This issue has been around for a while, we never figured out why some people have it and not others.

Probably how Windows assigns threads, background processes on startup or mainboard memory training for timings.
It's what I figured out from hundreds of hours of benching (not rpcs3). Probably wouldn't happen on Linux.

**Pull request merged**